Infer method

ag91 opened the pull request:
This attempts to increase the capabilities of the "Create New Symbol" action to allow new method definition.

Supported only Scala 2 for cases shown in tests: tests/cross/src/test/scala/tests/pc/InsertInferredMethodSuite.scala

Related discussion and feature request: https://github.com/scalameta/metals-feature-requests/issues/298

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/codeactions/CreateNewSymbol.scala:
```diff
@@ -37,15 +37,40 @@ class CreateNewSymbol() extends CodeAction {
       )
     }
 
+    def createNewMethod(
```
You need to add:
```
override type CommandData = l.TextDocumentPositionParams
def command: Option[ActionCommand] = Some(ServerCommands.InsertInferredMethod)
```
somewhere above in the class

ag91 replied:
done in https://github.com/scalameta/metals/pull/6877/commits/12ec7f86a0245b74b3a1b5597d466250bcac58ec

That helped, but the real issue was that the `handleCommand` wasn't defined and inferMethod needs it to apply edits.

tgodzik replied:
Ach, yes I hoped it would get suggested by the compiler automatically. At least that was the idea when I added those fields in CodeAction.

Good catch!

tgodzik commented on mtags-interfaces/src/main/java/scala/meta/pc/PresentationCompiler.java:
```diff
@@ -157,6 +157,11 @@ public abstract CompletableFuture<List<AutoImportsResult>> autoImports(String na
+	/**
+	 * Return the missing method
+	 */
+	public abstract CompletableFuture<List<TextEdit>> insertInferredMethod(OffsetParams params);
```
Could we reuse the codeAction interface, it should be universal for all actions

ag91 replied:
is there an example by any chance? Otherwise I will pick this up at the end, after making sure the cases I handled in the tests work running interactively (not all do yet!)

tgodzik replied:
I was supposed to quickly create an example, but then figured we can refactor the old methods in https://github.com/scalameta/metals/pull/6881

It's more complicated for the older action, but for anything new(as soon as my PR is merged) it should be as easy as:
1. Add a new CodeActionId string
2. Implement maybeCodeActionId from CodeAction trait
3. Add handling of new id to codeAction method
4. Add id to list of supported ones
5. Use codeAction method from compilers in handleCommand

Feel free to work on the meat of the issue in the compiler, we can get back to it later or I can fix that myself.

ag91 replied:
right, I seem to have achieved what I wanted! Things are working now :sun_with_face:

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/codeactions/CreateNewSymbol.scala:
```diff
@@ -37,15 +37,40 @@ class CreateNewSymbol() extends CodeAction {
+            if name.head.isLower && params
+              .getRange()
+              .overlapsWith(diags.head.getRange()) =>
+          createNewMethod(diags.head, name)
```
Let's also check what version of Scala it is since Scala 3is not implemented

ag91 replied:
done in https://github.com/scalameta/metals/pull/6877/commits/12ec7f86a0245b74b3a1b5597d466250bcac58ec

ag91 commented:
@tgodzik I have managed to handle the issues I experienced by running my feature interactively.
I am now tuning the action to allow edits in other documents via [WorkspaceEdit](https://github.com/ag91/metals/blob/infer-method/metals/src/main/scala/scala/meta/internal/rename/RenameProvider.scala#L127) (without inferring a method for classes wouldn't be that much useful... ). 

Meanwhile, I have a question:
wouldn't make sense to have a logging utility to let user report bugs with the action I am adding?
By that I mean this is what you get when something goes wrong with an action:
    ```
    SEVERE: Internal error: scala.concurrent.Future$$anon$2: Future.filter predicate is not satisfied
    java.util.concurrent.CompletionException: scala.concurrent.Future$$anon$2: Future.filter predicate is not satisfied
    Caused by: scala.concurrent.Future$$anon$2: Future.filter predicate is not satisfied
    ```
   
 but for debugging the issue I need the pprint.log of the tree and the position and possibly the original code.
 How would a user report an issue like this? Should I look into printing this information out?

In my experience I have had a bunch of issues with inferType and now I see why the metals.log didn't have any hints about what caused them (not necessarily the state of my project, but possibly because not all cases are handled xD)

ag91 commented:
@tgodzik also:
managed to convert to a WorkspaceEdit. In my testing adding a method to an imported object that exists in another file works ONLY IF the file is already open (I am testing with Emacs).
Not sure if that is a known limitation or there is a trick I have to use to tell the LSP client to open the file and then apply the edits?

And could I get a hint on how to pass the MiMa action? I can fix the scalafix one. What else is needed to merge this in?

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/Compilers.scala:
```diff
@@ -862,6 +863,35 @@ class Compilers(
+              .map {
+                case (uri, textEdits) => {
+                  (
+                    uri,
```
Maybe the uri here is wrong? Let's only return TextEdits maybe and use the uri from params like in ExtractMethodCodeAction?

ag91 replied:
the url here is correct, I found out it is an issue happening on restart of the server only (that is the way I test it).

tgodzik commented:
> And could I get a hint on how to pass the MiMa action? I can fix the scalafix one. What else is needed to merge this in?

We should be ok once we use the codeAction interface. I can update it all before merging

tgodzik commented:
I can't seem to make it work at all with `.method(..)` syntax, I can dig in later, but we could implement that later.

ag91 commented:
> I can't seem to make it work at all with `.method(..)` syntax, I can dig in later, but we could implement that later.

you were right, there was an issue with me not handling `NullaryMethodType`  (not sure why that didn't show in the tests): solved in https://github.com/scalameta/metals/pull/6877/commits/cc5b6e6cd4b6f38ef3621f0cd10fe1a8c13adae4

ag91 commented:
> > And could I get a hint on how to pass the MiMa action? I can fix the scalafix one. What else is needed to merge this in?
> 
> We should be ok once we use the codeAction interface. I can update it all before merging

Okay, I am removing the draft from this PR then.

tgodzik commented:
Is it ok if I push the changes related to the new codeAction inteface? That's now merged to main

tgodzik commented on mtags/src/main/scala-2/scala/meta/internal/pc/InferredMethodProvider.scala:
```diff
@@ -0,0 +1,520 @@
+   * @param t the enclosing template for the class/object/trait we are implementing.
+   */
+  private def inferEditPosition(t: Template): Position = {
+    // get text via reflection because Template could be in a different file
```
I would avoid handling separate files as it is quite complex to do. Anything dealing with multiple files gets complex pretty quickly.  The other file might be a library or a completely different module. You might be using a symbol that is not available in that different scope also. Or what happens if the file is Java? There is a lot of edge cases unfortunately. Keeping the functionality in a single file gives us much less complexity.

What about instead of going to another file we would create an implicit class providing the method if the symbol is not in the current file (that could be a separate improvement). And we could have a separate action that would export the implicit class to the actual type separately using semanticdb (anything dealing with multiple files usually uses semanticdb)

ag91 replied:
oh I didn't think about that, well spotted!
Without that it reduces the advantage of using the action, I think.
I am fine with removing this to retain Metals stability and I will look into baking an Emacs specific hack to support this same action to act on multiple files.
The implicit idea seems alright a workaround, although I wouldn't use it myself so we can leave it as a further work.

tgodzik replied:
I feel like this is still very useful, especially in current file. For anything outside the current context it might not be that predictable. You might have mistyped a method name and the action would create that mistyped in a totally different class, which you would not be that aware of.

But if you are very keen on this we could do it another way:
- find the definition path for the class/object symbol
- take type information about parameter types from semanticdb 
- invoke the inferredMethod on the file the type is from, give it that type data plus the name of the method and use presentation compiler in that file to create it.

That is more complex though unfortuntely.

tgodzik replied:
For this PR we could fist find definition and if it's the same file invoke the action otherwise not do that. This gives us the window to implement the second part at some time later on.

tgodzik replied:
It would also be worth seeing how other editors or language servers do it, but for now having a smaller, surer scope seems a way to go.

ag91 replied:
> For this PR we could fist find definition and if it's the same file invoke the action otherwise not do that. This gives us the window to implement the second part at some time later on.

indeed, I can just check the params.uri and the Template.symbol.source.uri match and print something like unsupported otherwise.

ag91 replied:
addressed in https://github.com/scalameta/metals/pull/6877/commits/a96e7ab4f844b4d0c9695bb263876ad231ad190a

tgodzik reviewed (changes requested):
We should tighten the scope of this PR I think, code action like this should really only work within a single file as otherwise it will become much more complex.

ag91 commented on mtags-interfaces/src/main/java/scala/meta/pc/PresentationCompiler.java:
```diff
@@ -158,211 +159,215 @@ public abstract CompletableFuture<List<AutoImportsResult>> autoImports(String na
 	public abstract CompletableFuture<List<TextEdit>> insertInferredType(OffsetParams params);
 
 	/**
-	 * Return the text edits for inlining a value.
```
oh, I think metals github action runs automatically before push (I had to apply to be able to push). Maybe you can try [difftastic](https://github.com/Wilfred/difftastic) or ignore [whitespaces in PR](https://stackoverflow.com/questions/37007300/how-to-ignore-whitespace-in-github-when-comparing)

tgodzik replied:
It's fine, but I think the formatting was done by the editor actually, but maybe we never run any formatting on it before :thinking: Would probably be cleaner to have that as a separate commit because otherwise any blame in the file will point at you change :sweat_smile:

tgodzik commented on mtags/src/main/scala-2/scala/meta/internal/pc/InferredMethodProvider.scala:
```diff
@@ -0,0 +1,496 @@
+             .asInstanceOf[NullaryMethodType]
+             .resultType
+         else containerSymbol.symbol.tpe) match {
+          case TypeRef(_, classSymbol, _) =>
```
Should we also make sure it's the same file here?

ag91 replied:
not necessary because the check is in `makeClassMethodEdits` which I made sure to call for each case that infers a method.

tgodzik commented on mtags/src/main/scala-2/scala/meta/internal/pc/InferredMethodProvider.scala:
```diff
@@ -0,0 +1,496 @@
+            // this gives me the position of the class
+            typedTreeAt(containerClass.symbol.pos) match {
+              // class case
+              case ClassDef(
```
Could we use ImplDef superclass for both ClassDef and ModuleDef?

ag91 replied:
it seems a good idea but a simple ClassDef -> ImplDef didn't compile.
I will look into the other issues you have highlighted since this feels minor and requires digging time.

tgodzik commented on mtags/src/main/scala-2/scala/meta/internal/pc/InferredMethodProvider.scala:
```diff
@@ -0,0 +1,496 @@
+   */
+  private def inferEditPosition(t: Template): Position = {
+    // get text via reflection because Template could be in a different file
+    val text = t.pos.source.content.map(_.toString).mkString
```
Since we only support single file we can just use text from the OffsetParams

tgodzik reviewed (changes requested):
Great work! I added a few comments.

There also seems to be an unhandled case:

`ObjectName.otherMethodWithoutParens` we do show the code action but fail to do anything. We should either not show the method (we can use trees for that and not show if we are in a select statement without and apply parent) or handle it in the compiler

tgodzik commented on mtags/src/main/scala-2/scala/meta/internal/pc/InferredMethodProvider.scala:
```diff
@@ -0,0 +1,496 @@
+                arguments
+              ) :: _ =>
+            makeEditsForListApply(arguments, containingMethod, errorMethod)
+          // List(1,2,3).map(myFn)
```
This and the one below does not work unfortunately, it's not really possible to get a type of List(1,2,3).map the current implementation will just produce:

def otherMethod(arg0: Int, arg1: Int, arg2: Int) = ???

which is not valid. We can't even know how many parameters it will have. And also this might be a much more complex expression.

tgodzik commented on mtags/src/main/scala-2/scala/meta/internal/pc/InferredMethodProvider.scala:
```diff
@@ -0,0 +1,496 @@
+              position = None
+            )
+          // List(1, 2, 3).map(nonExistent)
+          // List((1, 2)).map{case (x,y) => otherMethod(x,y)}
```
Is this actually the snippet this handles? Looks like this is a simple apply `containingMethod(arguments)` The same as the one above

I don't see any check for being inside another method.

ag91 replied:
yes, this is handling both List cases: both the non argument case, and the unpacking `case`

ag91 commented:
> I hope you are not too discouraged. This is a pretty hard one, which is why I probably abandoned this after the initial tries.
It is alright: it is a mix of I had more time before because I was on holiday and lost a bit of enthusiasm because the action will not be as useful as I needed it to be.

I am happy to wrap the PR up with your feedback though: I will reduce the cases and polish, possibly during the weekend.
I will get this in so Metals users can benefit ☀️

ag91 commented on tests/cross/src/test/scala/tests/pc/InsertInferredMethodSuite.scala:
```diff
@@ -0,0 +1,527 @@
+    """|trait Main {
+       |  def main() = {
+       |    val list = List(1, 2, 3)
+       |    def otherMethod(arg0: Int) = ???
```
I can create an issue later for an improvement, but now I would focus on wrapping this up instead of adding more features.
By the way: should I create multiple issues? That seems better so that the contribution can be small and easy (to invite other inexperienced contributors), in the sense of a "good first issue" tag kind of thing in github.

ag91 commented on mtags-interfaces/src/main/java/scala/meta/pc/PresentationCompiler.java:
```diff
@@ -156,6 +157,11 @@ public abstract CompletableFuture<List<AutoImportsResult>> autoImports(String na
+	/**
+	 * Return the missing method
+	 */
+	public abstract CompletableFuture<WorkspaceEdit> insertInferredMethod(OffsetParams params);
```
yes please, I haven't dived in the codeAction method sure that it would be easier for you to adapt.

ag91 commented on tests/cross/src/test/scala/tests/pc/InsertInferredMethodSuite.scala:
```diff
@@ -0,0 +1,527 @@
+    """|trait Main {
+       |  def main() = {
+       |    val list = List(1, 2, 3)
+       |    def otherMethod(arg0: Int, arg1: Int) = ???
```
same as https://github.com/scalameta/metals/pull/6877#discussion_r1835426877

ag91 commented on tests/cross/src/test/scala/tests/pc/InsertInferredMethodSuite.scala:
```diff
@@ -0,0 +1,527 @@
+       |
+       |""".stripMargin,
+    """|class X() {
+       |  def otherMethod(arg0: Boolean, arg1: String, arg2: Int) = ???
```
can be an improvement, I can create an issue for it as well if you like

ag91 commented on mtags/src/main/scala-2/scala/meta/internal/pc/InferredMethodProvider.scala:
```diff
@@ -0,0 +1,496 @@
+          .map(_.toString)
+          .mkString
+        if (params.uri().toString() != getPositionUri(insertPos)) {
+          println("You can only infer method for current file.")
```
good idea! Should I approach it like this for every spot where we return `Nil` as text edits?
My rationale is that in that case we will always fail weirdly, while it would be better to let the user know that is an unexpected case.
So I can create a utility to log/throw in all those places.

ag91 commented:
@tgodzik it turned out that handling the `Object.definition` was just adding a couple of cases
and the `List(1,2,3).map(myFn)` also seemed easy to patch up.
Hope we are closer at this point

tgodzik commented on mtags/src/main/scala-2/scala/meta/internal/pc/InferredMethodProvider.scala:
```diff
@@ -0,0 +1,520 @@
+      logger.warning(
+        s"This case ($c) is not currently supported by the infer-method code action."
+      )
+      throw new RuntimeException()
```
We can use `DisplayableException` similar to `inlineValue` does it.

tgodzik commented on mtags/src/main/scala-2/scala/meta/internal/pc/InferredMethodProvider.scala:
```diff
@@ -0,0 +1,520 @@
+            signature(
+              name = errorMethod.name.toString(),
+              paramsString =
+                Option(argumentsString(arguments.take(1)).getOrElse("")),
```
What if we have a method like:

```
object DoubleList{
  def apply(a : Int) = List( (a,b) )
}

class DoubList{
  def map(f: (Int, Int) => String) = ???
}


```

`DoubleList(1).map(nonExistent)` would create a wrong method. We can't know what is the type used in a method based on the arguments of the preceding apply.

tgodzik replied:
A very hacky way around it would be to create a function that take in everything like
`def myErrorFunction(a: Any) : Nothing` and with that rerun the presentation compiler to get the actual type of map. We do a similar thing with updating inferred type : https://github.com/scalameta/metals/blob/main/mtags/src/main/scala-2/scala/meta/internal/pc/InferredTypeProvider.scala#L89C9-L89C26

But maybe we could do it later.

ag91 replied:
ignored the cases, thanks

tgodzik commented on mtags/src/main/scala-2/scala/meta/internal/pc/InferredMethodProvider.scala:
```diff
@@ -0,0 +1,520 @@
+            unimplemented("other-than-method")
+        }
+      case _ =>
+        unimplemented("other-than-ident-or-select")
```
We could add better explanations here and use the DisplayableError which would then show up for the users.

ag91 replied:
this code is just a reference to find it back in code, I will make the logger into the DisplayableError.
I am unsure what would be a better explanation than saying "the case x is not currently supported for the action".
There could be many cases unsupported for these bits  (it would take time to list them and not sure how that would help a user who is likely just disappointed things don't work here)

tgodzik replied:
I think it's enough to make it into a sentence, so that it's clear what is not supported.

ag91 replied:
okay I applied the changes.

ag91 commented:
@tgodzik I think I have addressed all the comments. I think the last step is for you to address the MiMa issue and merge it when it is ready, no?
What a journey :sunny:

tgodzik commented:
> @tgodzik I think I have addressed all the comments. I think the last step is for you to address the MiMa issue and merge it when it is ready, no? What a journey ☀️

I think we are almost ready, I will add some final touches next week most likely related to the code actions interface. Thanks for your hard work!

tgodzik commented:
And looks like I broke some tests, I will take a look at them

tgodzik reviewed (approved):
We're green! Thanks for the work @ag91 !

I raised a follow up issue, we can fix things later.

ag91 commented:
Thank you @tgodzik  for all your explanations and the patch up of the PR: very grateful :bow: !
